### PR TITLE
feat: memoize `middleware` internally in `react-dom` and `react-native`

### PR DESCRIPTION
**Pull request by KubaJastrz**
It uses [Latest Ref Pattern](https://epicreact.dev/the-latest-ref-pattern-in-react/) to memoize `middlewares` array, so that users don't have to use `useMemo` themselves.

Affects `react-dom` and `react-native` packages.

- [x] add working test
- [x] ~~update docs~~
- [x] clean up code

### CODE REVIEW

**Comment by KubaJastrz** [comment]
@atomiks let me know if this is something you want to implement and I'll clean up the PR in the mean time.

I had to do some changes to run the tests as well.

I'm not really a big fan of this unit test, as the alternative of passing is literally an infinite loop (of `update` and component renders), so it might be slightly hard to debug if things break.

PS. `noUnusedLocals` + `ts-jest` is literally most annoying combination ever in terms of developer experience 😛

**Comment by atomiks** [comment]
Assuming it works it's great 😛  the `useMemo` thing is def annoying for the consumer so removing that requirement would be good.

**Comment by KubaJastrz** [comment]
@atomiks I see that `react-native` code has the same problem. I want to extract the latest ref hook to its own module, but not sure how to reuse that between the two workspaces.

Is it fine to duplicate that for now? Or do you have any better ideas?

**Comment by atomiks** [comment]
Duplicating is okay for now. We could make a non-published `utils` package that Rollup bundles in with React bundle (not external), but it doesn't matter that much for now tbh

**Comment by atomiks** [comment]
Regarding the docs, you should make a separate PR for that change specifically so we can wait until after publish to npm to merge

File: packages/react-native/src/utils/useLatestRef.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * @see https://epicreact.dev/the-latest-ref-pattern-in-react/
+ */
+export function useLatestRef<Value>(value: Value) {
+  const ref = useRef(value);
+
+  useIsomorphicLayoutEffect(() => {
+    ref.current = value;
+  });
+
```
Comment (atomiks) [inline packages/react-native/src/utils/useLatestRef.ts:13]: React Native should always `useLayoutEffect`, no need for this check (which will actually break it, I think).

Comment (KubaJastrz) [reply packages/react-native/src/utils/useLatestRef.ts:13]: I think the check is specifically designed not to break if `window` doesn't exist, but good catch.

**Pull request merged**